File: Week_04/Practicum/Reverse Linked List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* prev = dummy;

        for(int i = 0; i < left - 1; i++)
            prev = prev->next;
            
        ListNode *curr = prev->next;

        for(int i = 0; i < right - left; i++){
            ListNode *next = curr->next; 
            curr->next = next->next;
            next->next = prev->next;
            prev->next = next;
        }
        return dummy->next;
    }
};

File: Week_04/Practicum/Python/Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if head is None:
            return None
        if head.next is None:
            return head

        curr = head
        last_of_left_third = None
        idx = 1

        while idx < left:
            last_of_left_third = curr
            curr = curr.next
            idx += 1

        first_of_sorted_third = curr
        prev = None

        while idx <= right:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
            idx += 1

        if last_of_left_third:
            last_of_left_third.next = prev
        else:
            head = prev

        first_of_sorted_third.next = curr
        return head
